File: static_files/jslib/Diffie.js
// Simple implementation of DiffieHellman key exchange
// By: Hubert F. Espinola I


function getRandomPrime() {
	// local method array
	prime_list = [5n, 7n, 11n, 13n, 17n, 19n, 23n, 29n, 31n, 37n, 41n, 43n, 47n, 53n, 59n, 61n, 67n, 71n, 73n, 79n, 83n, 89n, 97n]
	random_index = Math.floor(Math.random() * prime_list.length)

	return prime_list[random_index]
}


// calculates big integer power with modulo n (if needed)
function bigIntPow(a, b, n=0n) {
	output = 1n
	for (var i = 0; i < b; i++)
		output *= a

	if (n != 0n)
		return output % n
	return output
}



class Diffie {
	constructor(Pvalue, QValue) {
		this.Pvalue  = Pvalue
		this.QValue  = QValue
		this.myValue = BigInt(Math.floor(Math.random() * 500))
	}

	getMixedVal() {
		return bigIntPow(this.Pvalue, this.myValue, this.QValue)
	}

	calculateKey(otherValue) {
		return bigIntPow(otherValue, this.myValue, this.QValue)
	}
}

File: modules/Diffie.py
# Simple implementation of diffie hellman key exchange
# By: Hubert F. Espinola I
import random

class Diffie:
	__prime_list = [27, 13, 19, 11]

	def __init__(self, PValue : int, QValue : int) -> None:
		self.PValue  = PValue
		self.QValue  = QValue
		self.myValue = random.randint(22222, 999999)

		self.myMixedVal    = None
		self.calculatedKey = None

		self.otherVal = None

	def getMixedVal(self):
		if (self.myMixedVal == None):
			self.myMixedVal = pow(self.PValue, self.myValue, self.QValue)
		return self.myMixedVal

	def calculateKey(self, othersMixedValue : int):
		if (self.calculatedKey == None):
			self.calculatedKey = pow(othersMixedValue, self.myMixedVal, self.QValue)		
		return self.calculatedKey

	def getCurrentKey(self):
		return self.otherVal

	@staticmethod
	def getRandomPrime():
		return random.choice(Diffie.__prime_list)

	@staticmethod
	def getCharHex(key : int):
		keyHex = hex(key).replace('0x', '')
		output = []

		if (len(keyHex) % 2 != 0):
			keyHex = '0' + keyHex

		for i in range(0, len(keyHex), 2):
			print(keyHex[i] + keyHex[i + 1])
			output.append(chr(int(keyHex[i] + keyHex[i + 1], base=16)))

		return output